make_count/make_count.tex

```diff
@@ -0,0 +1,293 @@
+\newtheorem{note}{Note}
+\newtheorem{hope}{Hope}
+\newtheorem{warning}{Warning}
+\newtheorem{problem}{Problem}
+\newtheorem{fear}{Fear}
+
+\begin{document}
+
+\maketitle
+
+\tableofcontents
+
+\section{MakeCount}
+
+\subsection{Implementation of MakeCount in Rust}
+\label{sub:rust-implementation}
+
+\inOpenDPRust{make\_count}{https://github.com/opendp/opendp/blob/main/rust/opendp/src/trans/count.rs}
+
+\permalink{https://github.com/opendp/opendp/blob/c3b5c3bd9fc50c556362b628f08c5fddea069b4d/rust/opendp/src/trans/count.rs\#L14-L27}{12/07/2021} (It is from \href{https://github.com/opendp/opendp/pull/173}{this pull request}.)
+The Rust code can also be seen below. 
+
+\begin{lstlisting}[language = Rust]
+pub fn make_count<TIA, TO>(
+) -> Fallible<Transformation<VectorDomain<AllDomain<TIA>>, AllDomain<TO>, SymmetricDistance, AbsoluteDistance<TO>>>
+    
+    where TO: ExactIntCast<usize> + One + DistanceConstant<IntDistance>, IntDistance: InfCast<TO> {
+    
+    Ok(Transformation::new(
+        VectorDomain::new_all(),
+        AllDomain::new(),
+        // think of this as: min(arg.len(), TO::MAX_CONSECUTIVE)
+        Function::new(move |arg: &Vec<TIA>|
+            TO::exact_int_cast(arg.len()).unwrap_or(TO::MAX_CONSECUTIVE)),
+        SymmetricDistance::default(),
+        AbsoluteDistance::default(),
+        StabilityRelation::new_from_constant(TO::one())))
+}
+\end{lstlisting}
+
+\subsection{Implementation of MakeCount in Python-style pseudocode, with preconditions}
+\label{sec:python-pseudocode}
+
+We now use Python-style pseudocode to present a representation of the Rust function.
+
+\emph{Recall that functions in the pseudocode are defined in the document \href{https://www.overleaf.com/project/60d215bf90b337ac02200a99}{``List of definitions used in the pseudocode"}.}
+
+\emph{The use of \texttt{code}-style parameters in the preconditions section below (for example, \texttt{input\_domain}) means that this information should be passed along to the \texttt{Transformation} constructor.}
+
+Here, we use preconditions to check for traits, and to specify the domains and metrics.
+
+\textbf{Preconditions}
+\begin{itemize}
+    \item \textbf{User-specified types:} The \texttt{make\_count} function takes two inputs: a generic input type \texttt{TIA} for the \texttt{Transformation} (meaning that the input vector to \texttt{Transformation} is of type \texttt{Vec(TIA)}), and a generic output type \texttt{TO} for the \texttt{Transformation}.
+    
+    \begin{itemize}
+        \item \texttt{TO} has traits \texttt{One}, \texttt{ExactIntCast(usize)}, and \texttt{DistanceConstant(IntDistance)}
+        
+        \item \texttt{IntDistance} has trait \texttt{InfCast(TO)}
+        
+        \item \question{The final bullet point above is not needed in this proof, but it is needed in the code so a hint can be constructed (otherwise a binary search would be needed to construct the hint). Should this precondition be included here or not?}

```

Comment:
Good question.  For now I'd keep it, but also keep your remark as a parenthetical.

Reply:
actually it looks like you do use it in the stability_relation below

Reply:
This is not used in the stability relation. We do use the fact that TO has trait InfCast(IntDistance). Although "TO has trait InfCast(IntDistance)" is not written in the code, because TO has trait DistanceConstant(IntDistance), we know then that TO has trait InfCast(IntDistance). (See the pseudocode definitions doc.)